File: src/main/scala/org/xbery/artbeams/products/repository/ProductMapper.scala
package org.xbery.artbeams.products.repository

import org.xbery.artbeams.common.assets.domain.AssetAttributes
import org.xbery.artbeams.common.assets.repository.AssetMapper
import org.xbery.artbeams.products.domain.Product
import org.xbery.overview.filter.Condition
import org.xbery.overview.mapper._
import org.xbery.overview.repo.Conditions

import java.util

/**
  * Maps {@link Product} entity to set of attributes and vice versa.
  * @author Radek Beran
  */
class ProductMapper() extends AssetMapper[Product, ProductFilter] {

  override protected def cls = classOf[Product]

  override val getTableName: String = "products"

  val slugAttr = add(Attr.ofString(cls, "slug").get(e => e.slug))
  val titleAttr = add(Attr.ofString(cls, "title").get(e => e.title))
  val fileNameAttr = add(Attr.ofString(cls, "filename").get(e => e.fileName.orNull))

  override def createEntity(attributeSource: AttributeSource, attributes: java.util.List[Attribute[Product, _]], aliasPrefix: String): Product = {
    val assetAttributes = createAssetAttributes(attributeSource, attributes.asInstanceOf[util.List[Attribute[_, _]]], aliasPrefix)
    Product(
      assetAttributes,
      slugAttr.getValueFromSource(attributeSource, aliasPrefix),
      titleAttr.getValueFromSource(attributeSource, aliasPrefix),
      Option(fileNameAttr.getValueFromSource(attributeSource, aliasPrefix))
    )
  }

  override def composeFilterConditions(filter: ProductFilter): util.List[Condition] = {
    val conditions = super.composeFilterConditions(filter)
    filter.slug.map(slug => conditions.add(Conditions.eq(this.slugAttr, slug)))
    filter.title.map(title => conditions.add(Conditions.eq(this.titleAttr, title)))
    conditions
  }

  override def entityWithCommonAttributes(entity: Product, common: AssetAttributes): Product = entity.copy(common = common)
}

object ProductMapper {
  lazy val Instance = new ProductMapper()
}

File: src/main/kotlin/org/xbery/artbeams/products/repository/ProductMapper.kt
package org.xbery.artbeams.products.repository

import org.xbery.artbeams.common.assets.domain.AssetAttributes
import org.xbery.artbeams.common.assets.repository.AssetMapper
import org.xbery.artbeams.products.domain.Product
import org.xbery.overview.filter.Condition
import org.xbery.overview.mapper.Attr
import org.xbery.overview.mapper.Attribute
import org.xbery.overview.mapper.AttributeSource
import org.xbery.overview.repo.Conditions

/**
 * Maps {@link Product} entity to set of attributes and vice versa.
 * @author Radek Beran
 */
open class ProductMapper : AssetMapper<Product, ProductFilter>() {
    override fun cls(): Class<Product> = Product::class.java
    override fun getTableName(): String = "products"

    val slugAttr: Attribute<Product, String> = add(Attr.ofString(cls(), "slug").get { e -> e.slug })
    val titleAttr: Attribute<Product, String> = add(Attr.ofString(cls(), "title").get { e -> e.title })
    val fileNameAttr: Attribute<Product, String> =
        add(Attr.ofString(cls(), "filename").get { e -> e.fileName })
    val confirmationMailingGroupIdAttr: Attribute<Product, String> =
        add(Attr.ofString(cls(), "confirmation_mailing_group_id").get { e -> e.confirmationMailingGroupId })
    val mailingGroupIdAttr: Attribute<Product, String> =
        add(Attr.ofString(cls(), "mailing_group_id").get { e -> e.mailingGroupId })

    override fun createEntity(
        attributeSource: AttributeSource,
        attributes: List<Attribute<Product, *>>,
        aliasPrefix: String?
    ): Product {
        val assetAttributes = createAssetAttributes(attributeSource, attributes as List<Attribute<*, *>>, aliasPrefix)
        return Product(
            assetAttributes,
            slugAttr.getValueFromSource(attributeSource, aliasPrefix ?: ""),
            titleAttr.getValueFromSource(attributeSource, aliasPrefix ?: ""),
            fileNameAttr.getValueFromSource(attributeSource, aliasPrefix ?: ""),
            confirmationMailingGroupIdAttr.getValueFromSource(attributeSource, aliasPrefix ?: ""),
            mailingGroupIdAttr.getValueFromSource(attributeSource, aliasPrefix ?: "")
        )
    }

    override fun composeFilterConditions(filter: ProductFilter): MutableList<Condition> {
        val conditions = super.composeFilterConditions(filter)
        filter.slug?.let { slug -> conditions.add(Conditions.eq(this.slugAttr, slug)) }
        filter.title?.let { title -> conditions.add(Conditions.eq(this.titleAttr, title)) }
        return conditions
    }

    override fun entityWithCommonAttributes(entity: Product, common: AssetAttributes): Product =
        entity.copy(common = common)

    companion object {
        val Instance: ProductMapper = ProductMapper()
    }
}
